### PR TITLE
chore: Post deployment check

### PR DESCRIPTION
**Pull request by sdstolworthy**
#### Description of changes

Adds a post-deployment check to ensure that binaries are available in GitHub, S3, and NPM.

See:
https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli/21319/workflows/ee734185-6b1a-47b3-b42d-46b4e063f76c
and
https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli/21318/workflows/077217db-c070-4e95-a87a-f21fda5e6cc2

for examples of how this will look in practice.

image
image

#### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: scripts/verify-deployment.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+type ProcessArgs = {
+  version: string;
+  isPrerelease: boolean;
+};
+const githubBinaries = ['amplify-pkg-linux-arm64.tgz', 'amplify-pkg-linux.tgz', 'amplify-pkg-macos.tgz', 'amplify-pkg-win.exe.tgz'];
+const parseArgs = async (): Promise<ProcessArgs> => {
+  const args = yargs(process.argv.slice(2))
+    .options({
+      v: { type: 'string', demandOption: true },
+      p: { type: 'boolean', default: false },
```
Comment (sobolk) [inline scripts/verify-deployment.ts:13]:
```suggestion
      version: { type: 'string', demandOption: true },
      prerelease: { type: 'boolean', default: false },
```

File: scripts/verify-deployment.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  const buildUrl = (binary: string) => `https://package.cli.amplify.aws/${version}/${binary}`;
+  const responses = await Promise.all(s3Binaries.map((b) => fetch(buildUrl(b))));
+  const responseStatuses = responses.map((r) => r.status);
+  return responseStatuses.every((s) => s === 200);
+};
+
+const main = async () => {
+  const { version, isPrerelease } = await parseArgs();
+  console.log(`#### Verifying version ${version} deployed correctly ####`);
+  const isGitHubSatisfied = isPrerelease || (await existsInGitHub(version));
```
Comment (sobolk) [inline scripts/verify-deployment.ts:48]: pre-release would have binaries too no ?
look at https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli/21287/workflows/ab5552e3-1bee-43ce-9406-1f54738d914e 
We upload binaries , then craft pre-release, ship to npm and promote pre-release to latest. All share same binaries and version number.

The `existsInGitHub` should take `isPrerelease` flag and check status of release (i.e. is latest or no)

File: .circleci/config.base.yml [hunk 3]
```diff
@@ -1325,6 +1345,27 @@ workflows:
                 - /run-e2e-with-rc\/.*/
                 - /tagged-release\/.*/
                 - /tagged-release-without-e2e-tests\/.*/
+      - deployment_verification_release:
```
Comment (sobolk) [inline .circleci/config.base.yml:1348]:
```suggestion
      - deployment_verification_post_release:
```

File: .circleci/config.base.yml [hunk 4]
```diff
@@ -1325,6 +1345,27 @@ workflows:
...
+          requires:
+            - publish_to_npm
+            - github_release
+            - upload_pkg_binaries
+          filters:
+            branches:
+              only:
+                - release
+      - deployment_verification_prerelease:
```
Comment (sobolk) [inline .circleci/config.base.yml:1357]:
```suggestion
      - deployment_verification_pre_release:
```

File: .circleci/config.base.yml [hunk 5]
```diff
@@ -1325,6 +1345,27 @@ workflows:
...
+          requires:
+            - publish_to_npm
+            - github_release
```
Comment (sobolk) [inline .circleci/config.base.yml:1360]:
```suggestion
            - github_prerelease
```
?

File: .circleci/config.base.yml [hunk 6]
```diff
@@ -1325,6 +1345,27 @@ workflows:
...
+            - upload_pkg_binaries
+          filters:
+            branches:
+              only:
+                - /release_rc\/.*/
+                - /run-e2e-with-rc\/.*/
+                - /tagged-release\/.*/
+                - /tagged-release-without-e2e-tests\/.*/
```
Comment (sobolk) [inline .circleci/config.base.yml:1368]: We don't publish to pre-release on github for tagged releases.

File: scripts/verify-deployment.ts (hunk 2 above)
Comment (sdstolworthy) [reply scripts/verify-deployment.ts:48]: Prerelease may not be the correct word here. I'm referring to RCs and tagged releases with this. Those do not get uploaded to GitHub.

File: .circleci/config.base.yml (hunk 6 above)
Comment (sdstolworthy) [reply .circleci/config.base.yml:1368]: I clarified the naming here. Tagged and RC releases do not get deployed to github. They need to be treated differently so the verification doesn't fail when checking github.

File: scripts/verify-deployment.ts (hunk 1 above)
Comment (sdstolworthy) [reply scripts/verify-deployment.ts]: Added aliases for the flags

File: .circleci/config.base.yml [hunk 7]
```diff
@@ -899,6 +899,26 @@ jobs:
...
+    <<: *linux-e2e-executor-large
+    steps:
+      - restore_cache:
+          key: amplify-cli-repo-{{ .Branch }}-{{ .Revision }}
+      - run:
+          name: Verify Release Deployment
+          command: |
+            version=$(cat .amplify-pkg-version)
+            yarn ts-node scripts/verify-deployment.ts -v $version
+  deployment_verification_rc_and_tagged:
```
Comment (sobolk) [inline .circleci/config.base.yml:912]:
```suggestion
  deployment_verification_rc_or_tagged:
```

File: .circleci/config.base.yml [hunk 8]
```diff
@@ -899,6 +899,26 @@ jobs:
...
+    <<: *linux-e2e-executor-large
+    steps:
+      - restore_cache:
+          key: amplify-cli-repo-{{ .Branch }}-{{ .Revision }}
+      - run:
+          name: Verify Tagged and RC Deployment
```
Comment (sobolk) [inline .circleci/config.base.yml:918]:
```suggestion
          name: Verify Tagged or RC Deployment
```

File: .circleci/config.base.yml [hunk 9]
```diff
@@ -899,6 +899,26 @@ jobs:
...
+          command: |
+            version=$(cat .amplify-pkg-version)
+            yarn ts-node scripts/verify-deployment.ts --version $version --prerelease true
```
Comment (sobolk) [inline .circleci/config.base.yml:921]: nit: `yargs` support so called boolean flags. I.e. the sole presence of flag indicates it's `true`.

File: .circleci/config.base.yml [hunk 10]
```diff
@@ -1325,6 +1345,27 @@ workflows:
...
+      - deployment_verification_post_release:
+          requires:
+            - publish_to_npm
+            - github_release
+            - upload_pkg_binaries
+          filters:
+            branches:
+              only:
+                - release
+      - deployment_verification_rc_and_tagged:
```
Comment (sobolk) [inline .circleci/config.base.yml:1357]:
```suggestion
      - deployment_verification_rc_or_tagged:
```

File: scripts/verify-deployment.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  version: string;
+  isPrerelease: boolean;
+};
+const githubBinaries = ['amplify-pkg-linux-arm64.tgz', 'amplify-pkg-linux.tgz', 'amplify-pkg-macos.tgz', 'amplify-pkg-win.exe.tgz'];
+const parseArgs = async (): Promise<ProcessArgs> => {
+  const args = yargs(process.argv.slice(2))
+    .version(false)
+    .options({
+      v: { alias: 'version', type: 'string', demandOption: true },
+      p: { alias: 'prerelease', type: 'boolean', default: false },
```
Comment (sobolk) [inline scripts/verify-deployment.ts:14]:
```suggestion
      gh: { alias: 'include-github', type: 'boolean', default: false },
```

To de-confuse naming.

**Review by sobolk — DISMISSED** [review]

**Review by lazpavel — DISMISSED** [review]

**Review by lazpavel — APPROVED** [review]

**Pull request merged**